mgr/telemetry: Add Ceph Telemetry module to send reports back to project

wido opened the pull request:
This Manager Module will send statistics and version information from
a Ceph cluster back to telemetry.ceph.com if the user has opted-in on sending
this information.

Additionally a user can tell that the information is allowed to be made
public which then allows other users to see this information.

Signed-off-by: Wido den Hollander <wido@42on.com>

tchaikov commented on src/brag/CMakeLists.txt:
```diff
@@ -1,3 +0,0 @@
-install(PROGRAMS
-  client/ceph-brag
-  DESTINATION bin)
```
```
CMake Error at src/CMakeLists.txt:1103 (add_subdirectory):
  add_subdirectory given source "brag" which is not an existing directory.
```
might want to update `src/CMakeLists.txt:1103` accordingly.

tchaikov commented on doc/mgr/telemetry.rst:
```diff
@@ -0,0 +1,36 @@
+Telemetry plugin
```
```
Warning, treated as error:
/home/jenkins-build/build/workspace/ceph-pr-docs/doc/mgr/telemetry.rst:document isn't included in any toctree
Build step 'Execute shell' marked build as failure
```
could you add this file to doc/mgr/index.rst

tchaikov commented on src/telemetry/server/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+flask
+flask_restful
```
add [elasticsearch](https://pypi.org/project/elasticsearch/)

tchaikov commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,334 @@
+from mgr_module import MgrModule
+
+try:
+    import requests
```
ceph-common depends on python-requests, so i guess we can assume that `requests` is always available?

liewegas commented:
```
Compiling /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/13.0.2-1664-g800d59c/rpm/el7/BUILDROOT/ceph-13.0.2-1664.g800d59c.el7.x86_64/usr/lib64/ceph/mgr/telemetry/module.py ...
  File "/usr/lib64/ceph/mgr/telemetry/module.py", line 83
    return str(string).lower() in ['true', 'yes', 'on']:
```

tchaikov commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,327 @@
+
+    @staticmethod
+    def str_to_bool(string):
+        return str(string).lower() in ['true', 'yes', 'on']:
```
as Sage commented, should remove the trailing `:`.

wido commented:
Thanks for the feedback @liewegas and @tchaikov, fixed!

jcsp commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,327 @@
+        'interval': 72
+    }
+
+    COMMANDS = [
```
Need to add an OPTIONS definition here (see message to list "mgr module interface change in mimic: config vs. store separation"), since the new mgr config stuff landed.

jcsp commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,327 @@
+        if self.last_upload is not None:
+            self.last_upload = int(self.last_upload)
+
+        self.report_id = self.get_config('report_id', None)
```
For internal stuff like this should use the new `get/set_store` functions

tchaikov commented:
```
error: File not found: /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/13.0.2-1793-gb7ecf21/rpm/el7/BUILDROOT/ceph-13.0.2-1793.gb7ecf21.el7.x86_64/usr/bin/ceph-brag


RPM build errors:
    File not found: /home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/13.0.2-1793-gb7ecf21/rpm/el7/BUILDROOT/ceph-13.0.2-1793.gb7ecf21.el7.x86_64/usr/bin/ceph-brag
```

@wido i think you need to update ceph.spec.in and debian/ceph-common.install accordingly.

tchaikov commented on ceph.spec.in:
```diff
@@ -1158,7 +1158,6 @@ fi
 %{_bindir}/rbd-replay-prep
 %endif
 %{_bindir}/ceph-post-file
-%{_bindir}/ceph-brag
```
https://github.com/ceph/ceph/pull/21501#issuecomment-384315789

please update debian packaging also.

tchaikov commented:
@wido also could you squash related change to a single one? i.e. one commit to remove brag, and another commit to add telemetry.

wido commented:
@tchaikov I rebased, fixed the DEB packaging and split it up into two commits.

I think everything is still there, but such rebases are always things which can easily go wrong :)

tchaikov commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,355 @@
+        mon_map = self.get('mon_map')
+        osd_map = self.get('osd_map')
+        fs_map = self.get('fs_map')
+        df = self.get('df')
```
could you dump some pool stats in the reports as well? as Sage suggested in the mailing list? and the service_map?

wido replied:
I'd have to look that up. At the moment I'm very limited on time so I'd rather have this module in before the Mimic release so that it's at least up there.

I will look into this later.

tchaikov commented on src/telemetry/README.md:
```diff
@@ -0,0 +1,4 @@
+# Ceph-telemetry
+This is the renewed version of `ceph-brag` where clusters running the `telemetry` module in their Manager can send back information about the cluster back to the project.
+
+This does not send any object data, it only sends statistics and information about the cluster.
```
could you note that Elasticsearch is prerequisite of ceph-telemetry here?

tchaikov commented on src/telemetry/server/ceph_telemetry/rest/index.py:
```diff
@@ -0,0 +1,7 @@
+
+class Index(Resource):
+    def get(self):
+        return jsonify(status=True)
```
i don't really understand how this helps us to access the collected stats?

wido replied:
It is a thing of me to make sure it responds on a GET so that you can do a health check with a external service to see if the webservice is alive.

wido commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,355 @@
+                try:
+                    self.last_report = self.compile_report()
+                except:
+                    self.log.exception('Exception while compiling report:')
```
Doesn't the self.log.execption do that already?

tchaikov commented:
@wido could you review the comments i posted yesterday? quite a few of them are not addressed.

wido commented:
@tchaikov I think I did? You had some documentation comments which I addressed by pushing/rebasing my commits.

The outstanding one is the pool stats and service map which I'm not able to work on at the moment.

tchaikov commented:
@wido to name some of them:

- https://github.com/ceph/ceph/pull/21501/commits/13945de6b1535382a883958aadd5edacccd51043#r185384160
- https://github.com/ceph/ceph/pull/21501/commits/13945de6b1535382a883958aadd5edacccd51043#r185385041
- https://github.com/ceph/ceph/pull/21501/commits/13945de6b1535382a883958aadd5edacccd51043#r185384337
- https://github.com/ceph/ceph/pull/21501/commits/13945de6b1535382a883958aadd5edacccd51043#r185384813
- https://github.com/ceph/ceph/pull/21501/commits/13945de6b1535382a883958aadd5edacccd51043#r185385958
- https://github.com/ceph/ceph/pull/21501/commits/13945de6b1535382a883958aadd5edacccd51043#r185385877
- https://github.com/ceph/ceph/pull/21501/commits/13945de6b1535382a883958aadd5edacccd51043#r185386605

jcsp commented:
Links aren't working for me either, but if I go to the "files changed" I can see comments that github is for some reason not showing on this page - some kind of glitch I guess.

jcsp commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,355 @@
+        if option == 'interval':
+            try:
+                value = int(value)
+            except (ValueError, TypeError):
```
int(None) will give TypeError -- not sure if there's a path that can pass None into this function but it seems a reasonable thing to catch

tchaikov replied:
but i checked the callers of `self.set_config_option()`, none of them passes None as value.

tchaikov replied:
@wido under what condition can we have `TypeError`?

wido replied:
If for some reason a None or empty strings knows to sneak in in some way. A user might provide this through the CLI

wido commented:
@tchaikov and @jcsp I have updated the code.

I took a quick glance at the service_map and that was easier than expected.

I also added the self-test to the module and added it to QA.

How does this look?

tchaikov commented on doc/mgr/telemetry.rst:
```diff
@@ -0,0 +1,36 @@
+================
+The telemetry plugin sends anonymous data about the cluster, in which it is running, back to the Ceph project.
+
+The data being send back to the project does not contain any sensitive data like pool names, object names, object contents or hostnames.
```
s/send/sent/

tchaikov commented on doc/mgr/telemetry.rst:
```diff
@@ -0,0 +1,36 @@
+
+The data being send back to the project does not contain any sensitive data like pool names, object names, object contents or hostnames.
+
+It contains counters and statistics on how the cluster has been deployed, the version of Ceph, the distribition it runs on and other parameters which help the project to gain a better understanding of the way Ceph is used.
```
might want to s/it/of the hosts/. as the second "it" should be indicating the same object as the first "It", but it is not.

tchaikov commented on doc/mgr/telemetry.rst:
```diff
@@ -0,0 +1,36 @@
+
+Interval
+--------
+The module compiles and sends a new report every 72 hours.
```
72 hours by default.

tchaikov commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,355 @@
+        },
+        {
+            "cmd": "telemetry show",
+            "desc": "Show last report or report to be send",
```
s/send/sent/

tchaikov commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,355 @@
+
+    @staticmethod
+    def is_valid_email(email):
+        regexp = "^.+@([?)[a-zA-Z0-9-.]+.([a-zA-Z]{2,3}|[0-9]{1,3})(]?))$"
```
is it possible a malicious client sends a **very** long email address?

tchaikov commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,355 @@
+        return metadata
+
+    def compile_report(self):
+        report = {'leaderboard': False, 'report_version': 1}
```
what are "leaderboard" and "report_version" for ?

tchaikov commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,376 @@
+
+        for option in ['description', 'contact', 'organization']:
+            report[option] = None
+            if self.config[option]:
```
nit, could be
```py
report[option] = self.config.get(option, None)
```

tchaikov commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,376 @@
+
+
+class Module(MgrModule):
+    run = False
```
i don't really understand why shall we make these variables class variable? as they are not shared among multiple instances of `Module`. and neither are they supposed to outlive the Module instance.

wido replied:
Well, that's a personal thing. I think it's easier to find where they have been defined. I could define them in the init() function, but that's not always very clear.

For me it doesn't mind that much where they are defined.

tchaikov commented on src/pybind/mgr/telemetry/module.py:
```diff
@@ -0,0 +1,355 @@
+        self.event.wait(10)
+
+        while self.run:
+            if not self.str_to_bool(self.config['enabled']):
```
i don't think it's necessary to convert the value to boolean. as `self.config['enabled']` is already a boolean when being added to this dictionary.

tchaikov commented on src/telemetry/README.md:
```diff
@@ -0,0 +1,4 @@
+# Ceph-telemetry
+This is the renewed version of `ceph-brag` where clusters running the `telemetry` module in their Manager can send back information about the cluster back to the project.
```
IMHO, this line is a little bit difficult to digest. how about:

> This is a renewed version of `ceph-brag`. It enables the Ceph project to collect statistics from users' clusters where the `telemetry` module is enabled, and index them in Elasticsearch.

wido commented:
@tchaikov Thanks for the comments! I saw them this time. I pushed new commits without a squash, I think that removed your last comments.

They should stay now to make it easier to review.

wido commented:
Done @tchaikov ! Squashed all commits. Looking forward to the merge :)

tchaikov commented:
@wido seems you rebased your change against latest master. could you rebase your change against mimic instead?
